modyn/supervisor/internal/pipeline_executor/evaluation_executor.py

```diff
@@ -0,0 +1,331 @@
+            return single_log
+
+        # As we are io bound by the evaluator server, GIL locking isn't a concern, so we can use multithreading.
+        with ThreadPoolExecutor(max_workers=num_workers or 1) as pool:
+            for eval_req in eval_requests:
+                task = partial(worker_func, eval_req)
+                tasks.append(pool.submit(task))
+
+            # join threads
+            for t in tasks:
+                logs.stage_runs.append(t.result())
+
+        return logs
+
+    def _single_evaluation(self, log: StageLog, eval_status_queue: Queue, eval_req: EvalRequest) -> None:
+        assert self.grpc.evaluator is not None, "Evaluator not initialized."
+        assert self.pipeline.evaluation
+        logger.info(
+            f"Evaluation Starts for model {eval_req.model_id} on split {eval_req.interval_start} "
+            f"to {eval_req.interval_end} of dataset {eval_req.dataset_id}."
+        )
+        dataset_config = next((d for d in self.pipeline.evaluation.datasets if d.dataset_id == eval_req.dataset_id))
+        log.info = SingleEvaluationInfo(eval_request=eval_req)
+        request = GRPCHandler.prepare_evaluation_request(
+            dataset_config.model_dump(by_alias=True),
+            eval_req.model_id,
+            self.pipeline.evaluation.device,
+            eval_req.interval_start,
+            eval_req.interval_end,
+        )
+        response: EvaluateModelResponse = self.grpc.evaluator.evaluate_model(request)
+        if not response.evaluation_started:
+            log.info.failure_reason = EvaluationAbortedReason.DESCRIPTOR.values_by_number[
+                response.eval_aborted_reason
+            ].name
+            logger.error(
+                f"Evaluation for model {eval_req.model_id} on split {eval_req.interval_start} to "
+                f"{eval_req.interval_end} not started with reason: {log.info.failure_reason}."
+            )
+            return
+
+        logger.info(
+            f"Evaluation started for model {eval_req.model_id} on split {eval_req.interval_start} "
+            f"to {eval_req.interval_end}."
+        )
+        reporter = EvaluationStatusReporter(
+            eval_status_queue,
+            response.evaluation_id,
+            eval_req.dataset_id,
+            response.dataset_size,
+        )
+        evaluation = {response.evaluation_id: reporter}
+        reporter.create_tracker()
+        self.grpc.wait_for_evaluation_completion(eval_req.training_id, evaluation)
+
+        eval_result_writer = JsonResultWriter(self.pipeline_id, eval_req.trigger_id, self.pipeline_logdir)
+        self.grpc.store_evaluation_results([eval_result_writer], evaluation)
+        self.grpc.cleanup_evaluations([int(i) for i in evaluation])
+        assert isinstance(eval_result_writer, JsonResultWriter)
+
+        log.info.results = (
+            eval_result_writer.results["datasets"][0][dataset_config.dataset_id]
```

Comment:
why the [0]?

Reply:
I think this has been like this before I touched it. One eval_result_writer is designed to handle multiple requests iirc, but we only use it for one.

Reply:
Ah i see. the whole result writer thing is a bit shaky anyways because there is duplication here and the pipeline log, but ok thanks for clearing that up!